File: trunk/delta3d/examples/testShaders/testShaders.cpp
#include <dtABC/application.h>
#include <dtCore/globals.h>
#include <dtCore/object.h>
#include <osg/Drawable>
#include <osg/PrimitiveSet>
#include <osg/Program>

using namespace dtABC;
using namespace dtCore;

class TestShadersApp : public Application
{

public:
   TestShadersApp( const std::string& configFilename = "testshadersconfig.xml" )
      : Application( configFilename )
   {
      mTotalTime = 0.0f;

      LoadGeometry();
      EnableShaders();

      Transform xform(0.0f, -3.0f, 0.0f, 0.0f, 0.0f, 0.0f);
      GetCamera()->SetTransform(&xform);

      GetWindow()->SetWindowTitle("testShaders");
   }

   void LoadGeometry()
   {
      mObject = new dtCore::Object("Happy Sphere");
      mObject->LoadFile("models/physics_happy_sphere.ive");
      AddDrawable(mObject.get());
   }

   void EnableShaders()
   {
      //create state set
      osg::StateSet* ss = mObject->GetOSGNode()->getOrCreateStateSet();

      //load the shader file
      mProg = new osg::Program;
      mDefault = new osg::Program;

      RefPtr<osg::Shader> vertexShader = new osg::Shader( osg::Shader::VERTEX );
      RefPtr<osg::Shader> fragmentShader = new osg::Shader( osg::Shader::FRAGMENT );

      mProg->addShader( vertexShader.get() );
      mProg->addShader( fragmentShader.get() );

      vertexShader->loadShaderSourceFromFile( GetDeltaRootPath()+ "/data/shaders/testshader.vert");
      fragmentShader->loadShaderSourceFromFile( GetDeltaRootPath()+ "/data/shaders/testshader.frag");

      ss->setAttributeAndModes( mProg.get(), osg::StateAttribute::ON );
      mEnabled = true;
   }

   virtual bool KeyPressed(const dtCore::Keyboard* keyboard, Producer::KeyboardKey key, Producer::KeyCharacter character)
   {
      bool verdict(false);
      if (key == Producer::Key_Escape)
      {
         this->Quit();
         verdict = true;
      }
      else if(key == Producer::Key_space)
      {
         osg::StateSet* ss = mObject->GetOSGNode()->getOrCreateStateSet(); 

         if(mEnabled)
         {
            ss->setAttributeAndModes( mProg.get(), osg::StateAttribute::OFF );
            ss->setAttributeAndModes( mDefault.get(), osg::StateAttribute::ON );
            mEnabled = false;
         }
         else
         {
            ss->setAttributeAndModes( mDefault.get(), osg::StateAttribute::OFF );
            ss->setAttributeAndModes( mProg.get(), osg::StateAttribute::ON );            
            mEnabled = true;
         }
         verdict = true;
      }

      return verdict;
   }

   virtual void PreFrame(const double deltaFrameTime )
   {
      mTotalTime += deltaFrameTime;

      osg::Matrix rotateMat;
      rotateMat.makeRotate(osg::DegreesToRadians(30.0f) * mTotalTime, osg::Vec3(0.0f, 0.0f, 1.0f));
      Transform xform;
      mObject->GetTransform(&xform);
      xform.SetRotation(rotateMat);
      mObject->SetTransform(&xform);
   }

   virtual ~TestShadersApp()
   {
   }

private:

   RefPtr<dtCore::Object>                    mObject;

   RefPtr<osg::Program>                      mProg; 
   RefPtr<osg::Program>                      mDefault;

   float                                     mTotalTime;
   bool                                      mEnabled;

   DeltaWin::Resolution                      mRes;
};

int main(int argc, char* argv[])
{

   SetDataFilePathList( GetDeltaRootPath() + "/examples/testShaders/;" +
                        GetDeltaDataPathList()  );

   RefPtr<TestShadersApp> app = new TestShadersApp( "testshadersconfig.xml" );
   app->Config();
   app->Run();

   return 0;
}

File: trunk/delta3d/examples/testShaders/testshaders.py
from PyDtUtil import *
from PyDtCore import *
from PyDtABC import *
from PyDtDAL import *


class TestShadersApp( Application ):
    def __init__(self, configFilename):
        Application.__init__(self, configFilename)
        self.mTotalTime = 0.0
        contextName = GetDeltaRootPath() + "/examples/data/demoMap"
        
        Project.GetInstance().SetContext( contextName )
        sm = ShaderManager.GetInstance()
        sm.LoadShaderDefinitions("Shaders/ShaderDefs.xml")
        self.LoadGeometry()
        self.EnableShaders()
        
        xform = Transform(0.0, -3.0, 0.0, 0.0, 0.0, 0.0)
        self.GetCamera().SetTransform(xform)

        self.GetWindow().SetWindowTitle("testShaders")
        
    def LoadGeometry(self):
        self.mObject = Object("Happy Sphere")
        self.mObject.LoadFile("models/physics_happy_sphere.ive")
        self.AddDrawable( self.mObject )  
   
    def EnableShaders(self):
        sm = ShaderManager.GetInstance()
        sp = sm.FindShaderPrototype("TestShader", "DemoShaders")
        if sp:
            osgn = self.mObject.GetOSGNode()
            sm.AssignShaderFromPrototype( sp, osgn )
            self.mEnabled = True
            
    def DisableShaders(self):
        sm = ShaderManager.GetInstance()
        osgn = self.mObject.GetOSGNode()
        sm.UnassignShaderFromNode( osgn )
        self.mEnabled = False
        
    def KeyPressed( self, keyboard, key ) :
        if key == int(KeyboardKey.KEY_Space):
            if self.mEnabled:
                self.DisableShaders()
            else:
                self.EnableShaders()
            return 1
        elif key == int(KeyboardKey.KEY_Escape):
            self.Quit()
            return 1
        return 0
        

dataPath = GetDeltaDataPathList()
SetDataFilePathList(dataPath + ";" + GetDeltaRootPath() + 
    "/examples/data" + ";" + GetDeltaRootPath() + "/examples/testShaders")

app = TestShadersApp( "testshadersconfig.xml" )
app.Config()
app.Run()
